python/tvm/relay/frontend/pytorch.py

```diff
@@ -127,8 +128,22 @@ def _is_quantized_tensor(data, prelude):
 # operator implementation
 def _elemwise(name):
     def _impl(inputs, input_types):
-        data0, data1 = _pytorch_promote_types(inputs[:2], input_types[:2])
-        return get_relay_op(name)(data0, data1)
+        dtype0, dtype1 = input_types[:2]
+        if isinstance(inputs[0], _expr.Expr):
+            dtype0 = _infer_type(inputs[0]).checked_type.dtype
+        if isinstance(inputs[1], _expr.Expr):
+            dtype1 = _infer_type(inputs[1]).checked_type.dtype
+
```

Comment:
This comes from weird behavior of ```prim::NumToTensor```. It converts int32 to int64 silently:
```
%11 : int = aten::size(%img.1, %10), scope: __module.model # /usr/local/lib/python3.6/dist-packages/torchvision/models/detection/generalized_rcnn.py:62:0
  %im_h : Long() = prim::NumToTensor(%11), scope: __module.model
```
Right now py frontend just follow use the same dtype for this op output. For an elemwise op, pytorch input dtype is ["int64", "int64"] which is fine. However, the actual input dtype is ["int64", "32"]. What I can do is to enhance ```_pytorch_promote_types``` so that we do _infer_type for every input and get actual input dtype, rather than solely relying on pytorch input dtype. Sounds like a plan?

Reply:
I think we would eventually want to look at using type propagation more.
However, the issue here is that PyTorch's default dtype for integral tensors is int64. I don't think we should be hacking around that, really, because we're bound to end up with cases where int64 is the right thing to have. If I understood the discussions on the forum correctly, the idea was to downcast 64 bit indexing to 32 based if it is considered safe.

Reply:
```%11 : int = aten::size(%img.1, %10)``` generates int32 but ```%im_h : Long() = prim::NumToTensor(%11)``` automatically converts it to int64, without any hint. When we converting ```prim::NumToTenso```, we can just follow the input type which is int32 here since there is no any other information. So this is about the weird behavior of ```prim::NumToTenso``` rather than indexing. I'm not sure how many other ops in pytorch has such behavior, but it looks like inferring actual input type in ```_pytorch_promote_types``` would fix these kind of issues.